java/src/com/ibm/streamsx/topology/generator/spl/Preprocessor.java

```diff
@@ -68,4 +71,45 @@ public void compositeColocateIdUsage(List<JsonObject> composites) {
         for (JsonObject composite : composites)
             pePlacementPreprocess.compositeColocateIdUse(composite);
     }
+
+    @SuppressWarnings("serial")
+    private void relocateHashAdders(){
+        final Set<JsonObject> hashAdders = new HashSet<>();
+        // Firstly, find each hashAdder
+        GraphUtilities.visitOnce(GraphUtilities.findStarts(graph),
+                new HashSet<BVirtualMarker>(),
+                graph,
+                new Consumer<JsonObject>(){
+                    public void accept(JsonObject op) {
+                        if(jstring(op, "kind").equals("com.ibm.streamsx.topology.functional.java::HashAdder")){
+                            hashAdders.add(op);
+                        }
+                    }
+        });
+
+        for(JsonObject hashAdder : hashAdders){
+            relocateHashAdder(hashAdder);
+        }
+    }
+
+    private void relocateHashAdder(JsonObject hashAdder){
+        // $Unparallel$ -> hashAdder -> $Parallel$ -> hashRemover
+        Set<JsonObject> parents = GraphUtilities.getUpstream(hashAdder, graph);
+        
+        // hashAdder has multiple parents
+        if (parents.size() != 1) return;
+
+        JsonObject parent = parents.iterator().next();
+
+        // hashAdder's parent has multiple children
```

Comment:
How could a hash adder have multiple downstream children?

Reply:
The **parent** of a `HashAdder` can have multiple children.

Reply:
I think I'm confused by the description of this PR which has:

> where HashAdder is $Unparallel's only child **and $parallel's only parent.**

So I thought the check was that the $parallel only has a single upstream operator, which is the hashadder, which will always be the case.

Reply:
Could you add comments as to why you are putting those restrictions in place, it's fine if it's because only the simple case is being tackled at the moment, but the code is much clearer if a reason is given.

Reply:
Sorry, I should have updated the PR description.